### PR TITLE
neox-2.x to master-2.x

### PR DESCRIPTION
**Pull request by KickSeason**
Apply `neox-2.x` to `master-2.x`.

I think it's ready now for `neox-2.x` to merge into `master-2.x`. 
So that we can arrange to deploy on `testnet` and `mainnet`.

@neo-project/core

### CODE REVIEW

**Comment by KickSeason** [comment]
@erikzhang @shargon @igormcoelho @vncoelho @Tommo-L 
Merge this?

**Comment by superboyiii** [comment]
These are pure changes on neox-2.x since neox-2.x are forked from master-2.x at the latest commit.

File: neo/ProtocolSettings.cs [hunk 1]
```diff
@@ -88,6 +89,7 @@ private ProtocolSettings(IConfigurationSection section)
             }
             this.SystemFee = sys_fee;
             this.SecondsPerBlock = section.GetValue("SecondsPerBlock", 15u);
+            this.StateRootEnableIndex = section.GetValue("StateRootEnableIndex", 0u);
```
Comment (shargon) [inline neo/ProtocolSettings.cs:92]: We should define by default the expected one

File: neo/Network/P2P/Payloads/StateRoot.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            Version = reader.ReadByte();
+            Index = reader.ReadUInt32();
+            PreHash = reader.ReadSerializable<UInt256>();
+            Root = reader.ReadSerializable<UInt256>();
+        }
+
+        public void Serialize(BinaryWriter writer)
+        {
+            ((IVerifiable)this).SerializeUnsigned(writer);
+            writer.Write(Witness is null ? new Witness[] { } : new Witness[] { Witness });
```
Comment (shargon) [inline neo/Network/P2P/Payloads/StateRoot.cs:71]:
```suggestion
            writer.Write(Witness is null ? Array.Empty<Witness>() : new Witness[] { Witness });
```

File: neo/Persistence/LevelDB/DbTrieStore.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+namespace Neo.Persistence.LevelDB
+{
+    public class DbTrieStore : IKVStore
+    {
+        private readonly DB db;
+        private readonly ReadOptions options;
+        private readonly WriteBatch batch;
+        private readonly byte prefix;
+
+        private readonly byte[] ROOT_KEY = Blockchain.GenesisBlock.Hash.ToArray();
```
Comment (shargon) [inline neo/Persistence/LevelDB/DbTrieStore.cs:15]: It's the same as https://github.com/neo-project/neo/pull/1803/files#diff-d9b8054e080c526156aa17afc3420f6cR14

File: neo/Network/RPC/RpcServer.cs [hunk 4]
```diff
@@ -749,5 +772,86 @@ private JObject ValidateAddress(string address)
             json["isvalid"] = scriptHash != null;
             return json;
         }
+
+        private JObject GetStateHeight()
+        {
+            var json = new JObject();
+            json["blockHeight"] = Blockchain.Singleton.Height;
```
Comment (shargon) [inline neo/Network/RPC/RpcServer.cs:779]: lowercase for neo3x compatibility?

File: neo/Consensus/ConsensusService.cs [hunk 5]
```diff
@@ -409,6 +421,15 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
                 return;
             }
 
+            byte[] stateRootHashData = context.MakeStateRoot().GetHashData();
```
Comment (vncoelho) [inline neo/Consensus/ConsensusService.cs:424]: Precise!

File: neo/Consensus/ConsensusService.cs [hunk 6]
```diff
@@ -424,9 +445,12 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
                         context.PreparationPayloads[i] = null;
             context.PreparationPayloads[payload.ValidatorIndex] = payload;
             byte[] hashData = context.MakeHeader().GetHashData();
+            byte[] stateData = context.MakeStateRoot().GetHashData();
```
Comment (vncoelho) [inline neo/Consensus/ConsensusService.cs:448]: We need to verify the StateRootSignature of the Primary, no?

File: neo/Consensus/RecoveryMessage.PreparationPayloadCompact.cs [hunk 7]
```diff
@@ -10,30 +11,45 @@ public class PreparationPayloadCompact : ISerializable
...
                 ValidatorIndex = reader.ReadUInt16();
                 InvocationScript = reader.ReadVarBytes(1024);
+                StateRootSignature = reader.ReadBytes(64);
             }
 
             public static PreparationPayloadCompact FromPayload(ConsensusPayload payload)
             {
+                byte[] StateRootSignature = Array.Empty<byte>();
+                ConsensusMessage message = payload.ConsensusMessage;
+                if (message is PrepareRequest req)
```
Comment (vncoelho) [inline neo/Consensus/RecoveryMessage.PreparationPayloadCompact.cs:32]: Very good as well.

File: neo/Network/RPC/RpcServer.cs [hunk 8]
```diff
@@ -749,5 +772,86 @@ private JObject ValidateAddress(string address)
...
+            json["stateHeight"] = Blockchain.Singleton.StateHeight;
+            return json;
+        }
+
+        private JObject GetStateRoot(JObject key)
```
Comment (vncoelho) [inline neo/Network/RPC/RpcServer.cs:784]: Maybe it would be better if it goes along with the `Height`, yes?

File: neo/neo.csproj [hunk 9]
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
     <Copyright>2015-2019 The Neo Project</Copyright>
     <AssemblyTitle>Neo</AssemblyTitle>
-    <Version>2.10.3</Version>
+    <Version>2.10.3-neox-preview1</Version>
```
Comment (vncoelho) [inline neo/neo.csproj:6]: if merged perhaps it will be something like 2.11.0, yes?

File: neo/Persistence/LevelDB/DbTrieStore.cs (hunk 3 above)
Comment (KickSeason) [reply neo/Persistence/LevelDB/DbTrieStore.cs:15]: Should be the same. Because they use the same store.

File: neo/ProtocolSettings.cs (hunk 1 above)
Comment (KickSeason) [reply neo/ProtocolSettings.cs:92]: OK

File: neo/Network/RPC/RpcServer.cs (hunk 4 above)
Comment (KickSeason) [reply neo/Network/RPC/RpcServer.cs:779]: Will do.

File: neo/Consensus/ConsensusService.cs (hunk 6 above)
Comment (KickSeason) [reply neo/Consensus/ConsensusService.cs:448]: Verified here:
https://github.com/neo-project/neo/blob/445c9900ad62ad2caf9ac5ad8ccbc7cc4e73d9d7/neo/Consensus/ConsensusService.cs#L424-L431

But didn't re-use `stateRootHashData`. Will fix.

File: neo/Network/RPC/RpcServer.cs (hunk 8 above)
Comment (KickSeason) [reply neo/Network/RPC/RpcServer.cs:784]: If `key` is `Number`, return state root of the height.

File: neo/neo.csproj (hunk 9 above)
Comment (KickSeason) [reply neo/neo.csproj:6]: Wil do.

File: neo/Network/RPC/RpcServer.cs (hunk 4 above)
Comment (KickSeason) [reply neo/Network/RPC/RpcServer.cs:779]: #1808

File: neo/Consensus/ConsensusService.cs (hunk 6 above)
Comment (KickSeason) [reply neo/Consensus/ConsensusService.cs:448]: #1809

File: neo/neo.csproj (hunk 9 above)
Comment (KickSeason) [reply neo/neo.csproj:6]: #1810

Comment (shargon) [reply neo/neo.csproj]: Merged

File: neo/Network/RPC/RpcServer.cs (hunk 4 above)
Comment (shargon) [reply neo/Network/RPC/RpcServer.cs]: Merged

File: neo/Persistence/LevelDB/DbTrieStore.cs (hunk 3 above)
Comment (shargon) [reply neo/Persistence/LevelDB/DbTrieStore.cs:15]: But it could be shared?

File: neo/ProtocolSettings.cs (hunk 1 above)
Comment (shargon) [reply neo/ProtocolSettings.cs:92]: Of course, only don't forget it.

File: neo/Consensus/ConsensusService.cs (hunk 6 above)
Comment (shargon) [reply neo/Consensus/ConsensusService.cs]: Merged

File: neo/Persistence/LevelDB/DbTrieStore.cs (hunk 3 above)
Comment (KickSeason) [reply neo/Persistence/LevelDB/DbTrieStore.cs:15]: Oh, I see.

File: neo/ProtocolSettings.cs (hunk 1 above)
Comment (KickSeason) [reply neo/ProtocolSettings.cs:92]: BTW, We have `StateRootEnableIndex` setting in `protocol.json`.
Besides, Height of `mainnet` is different from `testnet`. 
I don't think it's necessary to focus on this default value.

File: neo/Persistence/LevelDB/DbTrieStore.cs (hunk 3 above)
Comment (KickSeason) [reply neo/Persistence/LevelDB/DbTrieStore.cs:15]: Fixed in #1814

**Comment by KickSeason** [comment]
@neo-project/core Merge?

**Comment by shargon** [comment]
From my side I need another review before approve it, it's a big PR, also I think that we need more tests, @superboyiii could you test it deeply?

**Comment by KickSeason** [comment]
@shargon We have deployed on `testnet` for a long time except consensus optimization.

**Comment by superboyiii** [comment]
> From my side I need another review before approve it, it's a big PR, also I think that we need more tests, @superboyiii could you test it deeply?

I've tested this together with @KickSeason . It's already on testnet via last release expect consensus optimization.

**Comment by shargon** [comment]
It seems that testnet currently has a delayed https://neoscan-testnet.io/

**Comment by KickSeason** [comment]
> It seems that testnet currently has a delayed https://neoscan-testnet.io/

Testnet c# nodes stopped because of https://github.com/nspcc-dev/neo-go/issues/1270
We have 1 `neo-go` client in testnet consensus.

**Comment by superboyiii** [comment]
> It seems that testnet currently has a delayed https://neoscan-testnet.io/

We've proved that testnet issue is from [nspcc-dev/neo-go#1270](https://github.com/nspcc-dev/neo-go/issues/1270) since nspcc has a `neo-go` consensus node on testnet. And I've constantly send high pressure txs to test this version and it works fine till now.
image
image 
Please merge it to master-2.x, we plan to release new version on 10st August.

**Comment by roman-khimov** [comment]
> We've proved that testnet issue is from nspcc-dev/neo-go#1270

Yeah, we've already fixed that, but at the same time we have a situation where one malformed message knocked off all C# CNs from the network, I think there also is something to improve in C# code, especially if

> we plan to release new version on 10st August.

**Comment by superboyiii** [comment]
> > We've proved that testnet issue is from [nspcc-dev/neo-go#1270](https://github.com/nspcc-dev/neo-go/issues/1270)
> 
> Yeah, we've already fixed that, but at the same time we have a situation where one malformed message knocked off all C# CNs from the network, I think there also is something to improve in C# code, especially if
> 
> > we plan to release new version on 10st August.

Nice catch. You could maintain C# node first in next upgrade in testnet and let neo-go take place it whenever it's OK after test.

**Comment by KickSeason** [comment]
@shargon Merge?

File: neo/Consensus/ConsensusContext.cs [hunk 10]
```diff
@@ -231,15 +259,7 @@ public ConsensusPayload MakeRecoveryMessage()
...
-                prepareRequestMessage = new PrepareRequest
-                {
-                    ViewNumber = ViewNumber,
-                    TransactionHashes = TransactionHashes,
-                    Nonce = Nonce,
-                    NextConsensus = NextConsensus,
-                    MinerTransaction = (MinerTransaction)Transactions[TransactionHashes[0]],
-                    Timestamp = Timestamp
-                };
+                prepareRequestMessage = PreparationPayloads[PrimaryIndex].GetDeserializedMessage<PrepareRequest>();
```
Comment (shargon) [inline neo/Consensus/ConsensusContext.cs:262]: We should check if `PreparationPayloads[PrimaryIndex]` it's null

Comment (shargon) [reply neo/Consensus/ConsensusContext.cs:262]: Otherwise we should do the previous code

**Comment by shargon** [comment]
> @shargon Merge?

It is a very large pull request for only my decision, plase wait for @erikzhang review first

**Comment by KickSeason** [comment]
@erikzhang merge?

File: neo/Consensus/ConsensusContext.cs (hunk 10 above)
Comment (KickSeason) [reply neo/Consensus/ConsensusContext.cs:262]: We have this check 
https://github.com/neo-project/neo/blob/add411b88487da1eb0ea0f49d55be009487a4a22/neo/Consensus/ConsensusContext.cs#L260
if `TransactionHashes` is not `null`, it means node has received `PrepareRequest`.

**Pull request merged**